File: 0198-house-robber-打家劫舍/0198.py
class Solution:
    # 动态规划
    # 当前i金额最大值=(i-1金额最大值)or(i-2金额最大值+i金额)
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        moneyList = [nums[0]]
        if nums[1] > nums[0]:
            moneyList.append(nums[1])
        else:
            moneyList.append(nums[0])
        for i in range(2, len(nums)):
            robMoney = nums[i] + moneyList[i-2]
            if robMoney > moneyList[i-1]:
                moneyList.append(robMoney)
            else:
                moneyList.append(moneyList[i-1])
        print(moneyList)
        return moneyList[-1]

File: 0198-house-robber-打家劫舍/0198.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() < 1) return 0;
        if (nums.size() == 1) return nums[0];
        vector<int> money = {nums[0]};
        money.push_back(nums[0] > nums[1] ? nums[0] : nums[1]);
        for (int i = 2; i < nums.size(); ++ i) {
            int curr = nums[i] + money[i-2];
            if (money[i-1] > curr) curr = money[i-1];
            money.push_back(curr);
        }
        return money.back();
    }
};
/* 
// 与面试题 17.16. 按摩师相同
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        }
        if (nums.size() == 1) {
            return nums[0];
        }
        if (nums.size() == 2) {
            return max(nums[0], nums[1]);
        }
        std::vector<int> times = {nums[0], max(nums[0], nums[1])};
        for(int i = 2; i != nums.size(); ++i) {
            auto t = nums[i] + times[i-2] > times[i-1] ? nums[i] + times[i-2] : times[i-1];
            times.push_back(t);
        }
        return times.back();
    }
};
*/